File: typescript/design-patterns/proxy/proxy/ImageProxy.ts
// Same interface, defers the expensive part.
class ImageProxy implements Image {
  private real?: RealImage;
  constructor(private file: string) {}   // cheap

  display() {
    if (!this.real) {
      this.real = new RealImage(this.file);   // load ONCE, on first use
    }
    return this.real.display();              // then cache
  }
}

File: java/design-patterns/proxy/proxy/ImageProxy.java
class ImageProxy implements Image {
    private final String file;
    private RealImage real;     // null until first use
    ImageProxy(String file) { this.file = file; }

    public String display() {
        if (real == null) real = new RealImage(file);
        return real.display();
    }
}